### PR TITLE
Adding the xor16 hash function to bm_sim

### PR DESCRIPTION
**Pull request by Jithin-K-Sojan**
Added the xor16 hash to calculations.cpp.
Added a unit test for the same.

Tested it out with calls to both hash() and update_checksum().

### CODE REVIEW

**Comment by Jithin-K-Sojan** [comment]
Addresses issue #995 (HashAlgorithm - xor16)

File: src/bm_sim/calculations.cpp [hunk 1]
```diff
@@ -360,6 +360,22 @@ struct csum16 {
...
+    uint16_t mask = 0x00ff;
+    uint16_t final_xor_value = 0x0000;
+    /* Main loop - 2 bytes at a time */
+    for (unsigned int byte = 0; byte < len; byte += 2) {
+      uint16_t t1 = static_cast<uint16_t>(buf[byte]) << 8;
+      uint16_t t2 = 0x0000;
+      /* In case there is a 2nd byte */
+      if ((byte + 1) < len) t2 = static_cast<uint16_t>(buf[byte + 1]);
+      final_xor_value = final_xor_value ^ (t1 + (t2 & mask));
+    }
```
Comment (antoninbas) [inline src/bm_sim/calculations.cpp:374]: Even though this is a xor16 implementation, it's technically possible for len to be odd (e.g. 1, with a single-byte buffer), and you should handle this case (please add a test for this case as well).

**Review by antoninbas — CHANGES_REQUESTED** [review]

File: src/bm_sim/calculations.cpp (hunk 1 above)
Comment (Jithin-K-Sojan) [reply src/bm_sim/calculations.cpp]: Thanks for the comment.
The case of len being odd is handled. I have added a test for the same in a new commit.

File: src/bm_sim/calculations.cpp [hunk 2]
```diff
@@ -360,6 +360,22 @@ struct csum16 {
...
+struct xor16 {
+  uint16_t operator()(const char *buf, size_t len) const {
+    uint16_t mask = 0x00ff;
+    uint16_t final_xor_value = 0x0000;
+    /* Main loop - 2 bytes at a time */
+    for (unsigned int byte = 0; byte < len; byte += 2) {
+      uint16_t t1 = static_cast<uint16_t>(buf[byte]) << 8;
+      uint16_t t2 = 0x0000;
+      /* Condition returns false on the last byte of an odd-length sequence. */
+      if ((byte + 1) < len) t2 = static_cast<uint16_t>(buf[byte + 1]);
```
Comment (antoninbas) [inline src/bm_sim/calculations.cpp:372]: do we need to have this test in the loop body? There is no need to check whether we are at the end for every iteration.
You could write something like:
```cpp
while (len >= 2) {
  // ...
  len -= 2;
}
if (len > 0) {
  // handle last byte
}
```

Comment (Jithin-K-Sojan) [reply src/bm_sim/calculations.cpp]: The loop body has been changed. The check now happens only once after the loop.

**Review by antoninbas — COMMENTED** [review]
Code lgtm, but please address the style check failures reported by CI

**Comment by Jithin-K-Sojan** [comment]
So the style check failure seems to be coming from test/utils.cpp, the source of the error being the ordering of the header files.
I have resolved the error, but wanted to confirm if it is ok to add it to this PR since it is not related to the rest of the code in the PR.

**Comment by antoninbas** [comment]
@Jithin-K-Sojan the source of the failure is introduced by your PR:
```
/home/travis/build/p4lang/behavioral-model/tools/../src/bm_sim/calculations.cpp:368:  Missing space after ,  [whitespace/comma] [3]
```
I don't see another failure in the Travis logs.

**Comment by Jithin-K-Sojan** [comment]
Thanks @antoninbas the error you mentioned resolved the style check issue.

**Review by antoninbas — APPROVED** [review]
LGTM, thanks for contributing this

**Pull request merged**